### ISSUE TITLE
Ambiguous Typing for Default Lambda Arg when using Generics

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#7291]
I'm running into the following situation.

Let's say you have a function like this:

```python
I = TypeVar("I", bound=dict[str, Any] | Any)
V = TypeVar("V")


def transform_iterable(values: Iterable[I], func: Callable[[I], V] = lambda x: x) -> list[V]:
    return [func(value) for value in values]
```

Where you have some lambda, that might be overridden, or otherwise returns `I` if left alone.

Let's say you avoid passing in a custom callback.

```python
from typing import TypeVar, Any, Iterable, Callable, reveal_type

I = TypeVar("I", bound=dict[str, Any] | Any)
V = TypeVar("V")


def transform_iterable(values: Iterable[I], func: Callable[[I], V] = lambda x: x) -> list[V]:
    return [func(value) for value in values]

lst = my_func([1, 2, 3])
reveal_type(lst) # Revealed type is "list[I@my_func]". this should be list[int]
```

[Pyright Playground
](https://pyright-play.net/?strict=true&code=FAMwTg9gtgBALgTwA4EsB2BzGKpImOGAFWQFMA1AQzABoYBBNBOgSTlLEoCMAbUugMKUePbnzphSAN1LCA%2BoiSlgwFjAC8xMlTAAKAEQt9dLhACuaACbrLKAMZwA2gGc4tBkwC6MAD4eEAJTA5BpaSjoG5PpBKpakIPCcaM4g%2BFByKOycvKS6UsJmpM4AXDBsHGKkjiyedCAWdqVCIpWO1bUw5N6aolBclpQwAB6lQwEwALQAfDA8KK6OXcXAMKswknBmYGgwjvVodnkFpOOpYDD5PIXYO5eFzp4qAF6hbpTJZ%2BmZFTm6jgCMdAATHQAMyeIKSGTyRS5J4BIA)

Pyright is clever enough to understand that the type in the container (in this case a list) is actually `I`, but doesn't reason that it's an integer. Or perhaps we don't want a case like this to be supported for some reason, but in my mind if we can narrow to `I`, I don't see why we couldn't narrow even further to `int`, as it's provided for `values`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by max-muoto** [comment]
> Thanks for the bug report. This will be addressed in the next release.

Thanks Eric! Really appreciate it.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.351, which I just published.
